tests/Unit/GuzzleFileFetcherTest.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Tuf\Tests\Unit;
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Handler\MockHandler;
+use GuzzleHttp\HandlerStack;
+use GuzzleHttp\Middleware;
+use GuzzleHttp\Psr7\Response;
+use PHPUnit\Framework\TestCase;
+use Tuf\Client\GuzzleFileFetcher;
+use Tuf\Exception\DownloadSizeException;
+use Tuf\Exception\RepoFileNotFound;
+
+/**
+ * @coversDefaultClass \Tuf\Client\GuzzleFileFetcher
+ */
+class GuzzleFileFetcherTest extends TestCase
+{
+    /**
+     * Data provider for testFetchFile().
+     *
+     * @return array[]
+     *   Sets of arguments to pass to the test method.
+     */
+    public function dataProvider(): array
+    {
+        return [
+            [404, 256, RepoFileNotFound::class],
+            [403, 256, 'RuntimeException'],
+            [500, 256, 'RuntimeException'],
+            [200, 4, DownloadSizeException::class],
+            [200, 256, null],
+        ];
+    }
+
+    /**
+     * Tests various conditions when fetching a file.
+     *
+     * @param integer $statusCode
+     *   The response status code.
+     * @param integer $maxBytes
+     *   The maximum number of bytes to read from the response.
+     * @param string|null $exceptionClass
+     *   The expected exception class that will be thrown, if we expect the file
+     *   fetcher to throw an exception.
+     *
+     * @return void
+     *
+     * @dataProvider dataProvider
+     *
+     * @covers ::fetchFile
+     */
+    public function testFetchFile(int $statusCode, int $maxBytes, ?string $exceptionClass): void
+    {
+        $content = json_encode(['oolong', 'assam', 'matcha', 'herbal']);
+        $mockHandler = new MockHandler();
+        $mockHandler->append(new Response($statusCode, [], $content));
+        $handlerStack = HandlerStack::create($mockHandler);
+        // The httpErrors() middleware will throw an exception if the status
+        // code is not 200.
+        $handlerStack->push(Middleware::httpErrors());
+        $history = [];
+        $handlerStack->push(Middleware::history($history));
+        $client = new Client(['handler' => $handlerStack]);
+
+        $fetcher = new GuzzleFileFetcher($client);
+        if ($exceptionClass) {
+            $this->expectException($exceptionClass);
+        }
+        $data = $fetcher->fetchFile('test.json', $maxBytes);
+        $this->assertSame($content, $data);
+    }
+
+    /**
+     * Tests creating a file fetcher with a repo base URI.
+     *
+     * @return void
+     *
+     * @covers ::createFromUri
+     */
+    public function testCreateFromUri(): void
```

Comment:
Should we have a test for a call with no exception?

Reply:
We do. First line of `testCreateFromUri()`:
```
GuzzleFileFetcher::createFromUri('https://example.com');
```

Reply:
sorry read to fast. We don't assert the class of the return

Reply:
I meant to say lets assert the return value. like
`$this->assertInstanceOf(GuzzleFileFetcher::class, GuzzleFileFetcher::createFromUri('https://example.com'));`
otherwise we just know it didn't throw an exception. it still could return null or false.